### PR TITLE
Fix premature startup mode exit - issue 29944

### LINKED ISSUE #29944: cpuutil check in issue #28638 causes Thread Pool expansion limiting function to fail
**Describe the bug**  
When starting a Liberty server at FP 24.0.0.9 in CICS using the iFix from issue [28638](https://github.com/OpenLiberty/open-liberty/pull/28638) we fail to see the Thread pool held at 6 threads during the startup period. This appears to be because the Thread pool controller cpuutil check fails very quickly at startup due to a 0% CPU condition detected.

> [10/24/24, 13:10:01:373 GMT] 0000003a id=00000000 com.ibm.ws.threading.internal.ThreadPoolController           1      hang detected during startup, process cpu-util:  cpuUtil = 0.00%  - switching to normal controller operation

**Steps to Reproduce**  
Set following Thread pool size:
```
 <executor coreThreads="8" id="allowCICSconfigure" maxThreads="50"/>
```
Start a Liberty server and observe the number of threads active in the Liberty thread pool. In this scenario we see the thread pool reliably expand from 0 to 8 to 12 at startup.

**Expected behavior**  
The Thread pool should be limited to 6 threads until all apps have started. 

**Diagnostic information:**  
product = CICS LIBERTY 6.3.0, WAS FOR Z/OS 24.0.0.9 (wlp-1.0.93.cl241020240827-1743)
java.home = /java/java17_bit64_sr11_fp0/J17.0_64
java.version = 17.0.11
os = z/OS (02.05.00; s390x) (en_US)
In this environment there are 4 CPUs available, and a large number of Liberty features and 6 Java EE apps installed.

Trace of startup with this spec uploaded here [trace.zip](https://github.com/user-attachments/files/17508917/trace.zip)
`<logging consoleFormat="SIMPLE" traceSpecification="com.ibm.ws.threading.internal.ThreadPoolController=all=enabled"  maxFiles="10" maxFileSize="1024"/>`

### PR DESCRIPTION
**Pull request by gjdeval**
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

Fixes #29944

Two problems are addressed by these changes:
1) The first call to `CpuInfo.getJavaCpuUsage()` always returns 0, so we should not look at that first reading as an indicator that "no cpu was used in the prior thread pool controller cycle"
2) CICS replaces the original Liberty thread pool after Liberty launches, during the startup process, so watching the thread pool too closely immediately after launch can show unexpected behavior

### CODE REVIEW

File: dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java [hunk 1]
```diff
@@ -1423,9 +1433,12 @@ synchronized String evaluateInterval() {
         // we can't even think about adjusting the pool size until the underlying executor has aggressively
         // grown the pool to the coreThreads value, so if that hasn't happened yet we should just bail
         if (poolSize < coreThreads && initialStartupCompleted) {
-            return "poolSize < coreThreads";
+            // Try setting the pool to the correct size again
+            setPoolSize(coreThreads);
+            return "poolSize " + poolSize + " < coreThreads " + coreThreads;
         }
 
+        controllerCycle++;
```
Comment (jhanders34) [inline dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java]: Updating it here changes its definition from what is in the javadoc.  I think it is wrong to set it here based off of its current meaning.  Maybe you need a separate variable or maybe we check to see if the previous processCpuUtil is -1?  Also there is some other discussion regarding when cpu usage is called in a later comment which may make checking cycle count moot.

File: dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java [hunk 2]
```diff
@@ -1437,30 +1450,46 @@ synchronized String evaluateInterval() {
...
+                    // That's odd - let's try setting the pool size again
+                    setPoolSize(startupPoolSize);
+                    return "poolSize " + poolSize + " < startupPoolSize " + startupPoolSize;
                 }
-                if (deltaCompleted <= 0) {
-                    if (cpuUtil < startupHangCpuUtilThreshold) {
-                        if (tc.isEventEnabled()) {
-                            Tr.event(tc, "     hang detected during startup, process cpu-util: " + cpuUtilString + "%  - switching to normal controller operation");
-                        }
+                processCpuUtil = CpuInfo.getJavaCpuUsage();
```
Comment (jhanders34) [inline dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java]: When you said you were getting it earlier, I thought it meant that you were calling it before the initial run of the task, like in the constructor.  I would think that both the system and process cpu should be done at that time to prime them both.  Also when restoring from checkpoint I am guessing we should update them at the time as well.  Do not that you will want to rebase your branch on the latest code to pick up Tom's recent changes in this class.

Comment (jhanders34) [reply dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java]: I do wonder though for processCpuUtil and systemCpuUtil on whether both should be called above the try to always be calculated.

**Review by jhanders34 — CHANGES_REQUESTED** [review]

File: dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java (hunk 2 above)
Comment (gjdeval) [reply dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java]: I agree with both suggestions - had debated similar options without coming to a strong opinion.
So I will add calls to get process and system CPU to the constructor for early initialization, and move the calls for each cycle to before the try block.

File: dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java (hunk 1 above)
Comment (gjdeval) [reply dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java]: The primary purpose of `startupHangControllerCyclesSkip` is to allow for the CICS thread pool manipulation to complete before we start checking whether the thread pool has hung during startup, though I think there is no harm to letting the first couple of cycles go unchecked for other environments as well.

I'll move `controllerCycle++` back to its prior location, and just decrement  `startupHangControllerCyclesSkip` (have to remove `final` attribute) each time through the `if (!initialStartupCompleted) {` block until it is 0, at which point the logic for checking hang during startup can run.

**Comment by gjdeval** [comment]
#build

**Comment by contbld** [comment]
Your personal pipeline request is at https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?uuid=ceef6eb3-53b7-4104-8ad2-5444621ed16c

Target locations of links might be accessible only to IBM employees.

**Comment by contbld** [comment]
<h3>Code analysis and actions</h3>
DO NOT DELETE THIS COMMENT.

- 1 product code files were changed.

**Comment by contbld** [comment]
https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?uuid=ceef6eb3-53b7-4104-8ad2-5444621ed16c has errors/failures.

**Comment by gjdeval** [comment]
#libby

**Comment by contbld** [comment]
Your personal pipeline request is at https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?uuid=79cdaedc-988c-4971-8ed6-671f7ce0753e

Target locations of links might be accessible only to IBM employees.

**Comment by contbld** [comment]
<h3>Code analysis and actions</h3>
DO NOT DELETE THIS COMMENT.

- 1 product code files were changed.

**Comment by contbld** [comment]
https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?uuid=79cdaedc-988c-4971-8ed6-671f7ce0753e completed successfully!

**Review by jhanders34 — APPROVED** [review]

**Comment by gjdeval** [comment]
#libby

**Comment by contbld** [comment]
Your personal pipeline request is at https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?uuid=9aeb8fd9-7f67-4b1f-8fc0-ef44f749e234

Target locations of links might be accessible only to IBM employees.

**Comment by contbld** [comment]
<h3>Code analysis and actions</h3>
DO NOT DELETE THIS COMMENT.

- 1 product code files were changed.

**Comment by contbld** [comment]
https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?uuid=9aeb8fd9-7f67-4b1f-8fc0-ef44f749e234 completed successfully!

**Pull request merged**